bootloader/lkenv/lkenv.go

```diff
@@ -0,0 +1,406 @@
+	/* current kernel snap revision */
+	Snap_try_kernel [SNAP_NAME_MAX_LEN]byte
+
+	/* GADGET assets: current gadget assets revision */
+	Snap_gadget [SNAP_NAME_MAX_LEN]byte
+	/* GADGET assets: try gadget assets revision */
+	Snap_try_gadget [SNAP_NAME_MAX_LEN]byte
+
+	/**
+	 * Reboot reason
+	 * optional parameter to signal bootloader alternative reboot reasons
+	 * e.g. recovery/factory-reset/boot asset update
+	 */
+	Reboot_reason [SNAP_NAME_MAX_LEN]byte
+
+	/**
+	 * Matrix for mapping of boot img partion to installed kernel snap revision
+	 *
+	 * First column represents boot image partition label (e.g. boot_a,boot_b )
+	 *   value are static and should be populated at gadget built time
+	 *   or latest at image build time. Values are not further altered at run time.
+	 * Second column represents name currently installed kernel snap
+	 *   e.g. pi2-kernel_123.snap
+	 * initial value representing initial kernel snap revision
+	 *   is pupulated at image build time by snapd
+	 *
+	 * There are two rows in the matrix, representing current and previous kernel revision
+	 * following describes how this matrix should be modified at different stages:
+	 *  - at image build time:
+	 *    - extracted kernel snap revision name should be filled
+	 *      into free slow (first row, second row)
+	 *  - snapd:
+	 *    - when new kernel snap revision is being installed, snapd cycles through
+	 *      matrix to find unused 'boot slot' to be used for new kernel snap revision
+	 *      from free slot, first column represents partition label to which kernel
+	 *      snap boot image should be extracted. Second column is then populated with
+	 *      kernel snap revision name.
+	 *    - snap_mode, snap_try_kernel, snap_try_core behaves same way as with u-boot
+	 *  - bootloader:
+	 *    - bootloader reads snap_mode to determine if snap_kernel or snap_kernel is used
+	 *      to get kernel snap revision name
+	 *      kernel snap revision is then used to search matrix to determine
+	 *      partition label to be used for current boot
+	 *    - bootloader NEVER alters this matrix values
+	 *
+	 * [ <bootimg 1 part label> ] [ <kernel snap revison installed in this boot partition> ]
+	 * [ <bootimg 2 part label> ] [ <kernel snap revison installed in this boot partition> ]
+	 */
+	Bootimg_matrix [SNAP_BOOTIMG_PART_NUM][2][SNAP_NAME_MAX_LEN]byte
+
+	/**
+	 * name of the boot image from kernel snap to be used for extraction
+	 * when not defined or empty, default boot.img will be used
+	 */
+	Bootimg_file_name [SNAP_NAME_MAX_LEN]byte
+
+	/**
+	 * Gadget assets: Matrix for mapping of gadget asset partions
+	 * Optional boot asset tracking, based on bootloader support
+	 * Some boot chains support A/B boot assets for increased robustness
+	 * example being A/B TrustExecutionEnvironment
+	 * This matrix can be used to track current and try boot assets for
+	 * robust updates
+	 *
+	 * [ <boot assets 1 part label> ] [ <currently installed assets revison> ]
```

Comment:
typo "revison" , also these should say "in this partition" as well ?

Reply:
typo fixed and "in this partition" added